File: rust/herb-analysis/src/ruby_locals_index/offset_table.rs
use herb::location::Location;

use super::named_reference::NamedReference;

pub struct OffsetTable {
  line_starts: Vec<usize>,
}

impl OffsetTable {
  pub fn new(source: &str) -> Self {
    let mut line_starts = vec![0];

    for (index, byte) in source.bytes().enumerate() {
      if byte == b'\n' {
        line_starts.push(index + 1);
      }
    }

    Self { line_starts }
  }

  pub fn location_for(&self, reference: &NamedReference) -> Location {
    let (start_line, start_column) = self.position_at(reference.start_offset);
    let (end_line, end_column) = self.position_at(reference.start_offset + reference.length);

    Location::from(start_line as u32, start_column as u32, end_line as u32, end_column as u32)
  }

  pub fn position_at(&self, offset: usize) -> (usize, usize) {
    let following = self.line_starts.partition_point(|start| *start <= offset);
    let index = following.saturating_sub(1);
    let start = self.line_starts.get(index).copied().unwrap_or(0);

    (index + 1, offset - start)
  }
}

File: lib/herb/analysis/ruby_locals_index/offset_table.rb
# frozen_string_literal: true

module Herb
  module Analysis
    class RubyLocalsIndex
      class OffsetTable
        # @rbs!
        #   @line_starts: Array[Integer]

        #: (String) -> void
        def initialize(source)
          @line_starts = [0]

          source.each_byte.with_index do |byte, index|
            @line_starts << (index + 1) if byte == 0x0A
          end
        end

        #: (NamedReference) -> Herb::Location
        def location_for(reference)
          start_line, start_column = position_at(reference.start_offset)
          end_line, end_column = position_at(reference.start_offset + reference.length)

          Location.from(start_line, start_column, end_line, end_column)
        end

        #: (Integer) -> [Integer, Integer]
        def position_at(offset)
          following = @line_starts.bsearch_index { |start| start > offset }
          index = following ? following - 1 : @line_starts.length - 1
          start = @line_starts[index] || 0

          [index + 1, offset - start]
        end
      end
    end
  end
end
